lib/mix/lib/mix/utils.ex

```diff
@@ -498,7 +500,13 @@ defmodule Mix.Utils do
   def read_path(path, opts \\ []) do
     cond do
       url?(path) ->
-        read_httpc(path) |> checksum(opts)
+        task = Task.async(fn -> read_httpc(path) |> checksum(opts) end)
+        timeout = Keyword.get(opts, :timeout, 60_000)
+
+        case Task.yield(task, timeout) || Task.shutdown(task, :brutal_kill) do
+          {:ok, result} -> result
+          _ -> {:remote, :timeout}
```

Comment:
The return here needs to be a string. Maybe we can say `{:remote, "request timed out after #{timeout}ms"}`?